tracing/logger.go

```diff
@@ -0,0 +1,250 @@
+//tracing service.
+// The sending stops once Logger.Close is called.
+func (l *TraceLogger) AddStats(c *onet.Context, repeat time.Duration) {
+	go func() {
+		for {
+			select {
+			case <-time.After(repeat):
+				// Create a new trace that points to a dummy stackEntry,
+				//so the status can be sent to the service.
+				t, _ := l.newTrace(context.TODO(), "",
+					stackEntry{pkgPath: "go.dedis.ch/onet/v3/honeycomb",
+						method: "stats"})
+				for serName, status := range c.ReportStatus() {
+					t.hcTrace.AddField("status_"+serName, status)
+				}
+				t.send()
+			case <-l.done:
+				return
+			}
+		}
+	}()
+}
+
+// AddEnvironment reads the environment variables defined to initialize the
+// variables.
+// The following environmental variables are available:
+//   - TRACING_PRINT_SINGLE_SPAN - output a stack trace of single spans to
+//  let you trace where you could/should add a TRACING_ENTRY_POINT
+//   - TRACING_CREATE_SINGLE_SPANS -
+//   - TRACING_ENTRY_POINTS - a "::" separated list of entry points that can
+//  be used to refine the tracing.
+//  The name of the entry points are the same as given by
+//  TRACING_PRINT_SINGLE_SPAN
+//   - TRACING_DONE_MSGS - a "::" separated list of done msgs that will close
+//  the started trace. This is due to the simulation of actual traces,
+//  because we can't really know when the trace should end.
+func (l *TraceLogger) AddEnvironment() {
+	if envPSS := os.Getenv("TRACING_PRINT_SINGLE_SPAN"); envPSS != "" {
+		var err error
+		l.PrintSingleSpans, err = strconv.Atoi(envPSS)
+		if err != nil {
+			log.Warn("TRACING_PRINT_SINGLE_SPAN has non-integer value")
+		}
+	}
+	l.NoSingleSpans = os.Getenv("TRACING_CREATE_SINGLE_SPANS") == "true"
+	l.AddEntryPoints(strings.Split(os.Getenv("TRACING_ENTRY_POINTS"), "::")...)
+	l.AddDoneMsgs(strings.Split(os.Getenv("TRACING_DONE_MSGS"), "::")...)
+}
+
+func (l *TraceLogger) getTraceSpan(lvl int, msg string, callLvl int) (*traceWrapper,
+	*spanWrapper) {
+	l.logMutex.Lock()
+	defer l.logMutex.Unlock()
+	ses := parseLogs(strings.Split(log.Stack(), "\n")[1:])
+	ses = ses[0 : len(ses)-callLvl]
+	for i, se := range ses {
+		tr, ok := l.traces[se.methodStr()]
+		if ok {
+			for _, done := range l.doneMsgs {
+				if strings.Contains(msg, done) {
+					delete(l.traces, se.methodStr())
+					tr.send()
+					return nil, nil
+				}
+			}
+			sw := tr.stackToSpan(ses[i:])
+			sw.log(lvl, msg)
+			return tr, sw
+		}
+
+		if se.checkEntryPoint(l.entryPoints) != "" {
+			var child *spanWrapper
+			l.traces[se.methodStr()], child =
+				l.newTrace(context.TODO(), "", ses[i:]...)
+			child.log(lvl, msg)
+			return l.traces[se.methodStr()], child
+		}
+	}
+	if l.PrintSingleSpans > 0 {
+		fmt.Printf("Creating single trace for '%s' from:\n",
+			msg)
+		for i, se := range ses {
+			if i >= l.PrintSingleSpans {
+				break
+			}
+			fmt.Println("\t", se.pkgMethod())
+		}
+	}
+	if l.NoSingleSpans == false {
```

Comment:
```suggestion
	if !l.NoSingleSpans {
```

Reply:
I put this on purpose on `== false`, because the `No` is an inversion... But if you think it doesn't help read the code, I'll use the `!`, which, btw, is very similar to the `l`, and so might be overlooked...

Reply:
We can have both, by using a `logger` instead of `l`.